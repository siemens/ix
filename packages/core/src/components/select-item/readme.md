<!--
SPDX-FileCopyrightText: 2022 Siemens AG

SPDX-License-Identifier: MIT
-->

# ix-select-item

<!-- Auto Generated Below -->

## Properties

| Property             | Attribute  | Description                   | Type      | Default     |
| -------------------- | ---------- | ----------------------------- | --------- | ----------- |
| `hover`              | `hover`    | **_Internal_**                | `boolean` | `false`     |
| `label`              | `label`    | Displayed name of the item    | `string`  | `undefined` |
| `selected`           | `selected` | Whether the item is selected. | `boolean` | `false`     |
| `value` _(required)_ | `value`    | Item value                    | `any`     | `undefined` |

## Events

| Event       | Description  | Type                  |
| ----------- | ------------ | --------------------- |
| `itemClick` | Item clicked | `CustomEvent<string>` |

## Methods

### `onItemClick(event?: CustomEvent<HTMLIxDropdownItemElement>) => Promise<void>`

Internal

#### Returns

Type: `Promise<void>`

---
