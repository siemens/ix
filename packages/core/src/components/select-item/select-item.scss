/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

:host {
  display: block;
  position: relative;

  > ix-dropdown-item {
    width: 100%;
  }
}
