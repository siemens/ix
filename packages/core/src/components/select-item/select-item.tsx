/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  Method,
  Prop,
  Watch,
} from '@stencil/core';
import { IxSelectItemLabelChangeEvent } from './events';

@Component({
  tag: 'ix-select-item',
  styleUrl: 'select-item.scss',
  scoped: true,
})
export class SelectItem {
  @Element() hostElement: HTMLIxSelectItemElement;

  /**
   * Displayed name of the item
   */
  @Prop({ reflect: true }) label: string;

  /**
   * Item value
   */
  @Prop({ reflect: true }) value!: any;

  /**
   * Whether the item is selected.
   */
  @Prop() selected = false;

  /**
   * @internal
   */
  @Prop() hover = false;

  /**
   * Item clicked
   */
  @Event() itemClick: EventEmitter<string>;

  /**
   * @internal
   * @param event
   */
  @Method()
  async onItemClick(event?: CustomEvent<HTMLIxDropdownItemElement>) {
    event?.preventDefault();
    event?.stopPropagation();

    this.itemClick.emit(this.value);
  }

  componentDidRender() {
    if (!this.value) {
      throw Error('ix-select-item must have a `value` property');
    }
  }

  @Watch('label')
  labelChange(newValue: string, oldValue: string) {
    this.hostElement.dispatchEvent(
      new IxSelectItemLabelChangeEvent({
        newValue: newValue,
        oldValue: oldValue,
      })
    );
  }

  render() {
    return (
      <Host>
        <ix-dropdown-item
          checked={this.selected}
          label={this.label ? this.label : this.value}
          onItemClick={(e) => this.onItemClick(e)}
        ></ix-dropdown-item>
      </Host>
    );
  }
}
