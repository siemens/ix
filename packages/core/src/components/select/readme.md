<!--
SPDX-FileCopyrightText: 2022 Siemens AG

SPDX-License-Identifier: MIT
-->

# ix-select

<!-- Auto Generated Below -->

## Properties

| Property                  | Attribute                     | Description                                  | Type                     | Default                     |
| ------------------------- | ----------------------------- | -------------------------------------------- | ------------------------ | --------------------------- |
| `allowClear`              | `allow-clear`                 | Show clear button                            | `boolean`                | `false`                     |
| `disabled`                | `disabled`                    | If true the select will be in disabled state | `boolean`                | `false`                     |
| `editable`                | `editable`                    | Select is extendable                         | `boolean`                | `false`                     |
| `i18nPlaceholder`         | `i-1-8n-placeholder`          | Input field placeholder                      | `string`                 | `'Select an option'`        |
| `i18nPlaceholderEditable` | `i-1-8n-placeholder-editable` | Input field placeholder for editable select  | `string`                 | `'Type of select option'`   |
| `i18nSelectListHeader`    | `i-1-8n-select-list-header`   | Select list header                           | `string`                 | `'Please select an option'` |
| `mode`                    | `mode`                        | Selection mode                               | `"multiple" \| "single"` | `'single'`                  |
| `readonly`                | `readonly`                    | If true the select will be in readonly mode  | `boolean`                | `false`                     |
| `selectedIndices`         | `selected-indices`            | Indices of selected items                    | `string \| string[]`     | `[]`                        |

## Events

| Event                 | Description             | Type                              |
| --------------------- | ----------------------- | --------------------------------- |
| `addItem`             | Item added to selection | `CustomEvent<string>`             |
| `itemSelectionChange` | Item selection changed  | `CustomEvent<string \| string[]>` |

---
