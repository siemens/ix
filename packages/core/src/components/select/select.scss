/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@import 'common-variables';
@import 'mixins/fonts';
@import 'mixins/hover';
@import 'mixins/text-truncation';

:host {
  display: inline-block;
  position: relative;
  height: auto;
  border-radius: var(--theme-input--border-radius);

  .select {
    position: relative;
    display: flex;
    align-items: center;
    height: auto;
    background-color: var(--theme-input--background);
    border: var(--theme-input--border-thickness) solid
      var(--theme-input--border-color);
    border-radius: var(--theme-input--border-radius);
    padding: 0 0 0 $small-space;
    box-shadow: var(--theme-inset-shadow-1) !important;

    @include hover {
      background-color: var(--theme-input--background--hover);
      border-color: var(--theme-input--border-color--hover);
    }

    &:focus-within {
      background-color: var(--theme-input--background--focus);
      border-color: var(--theme-input--border-color--focus);
      outline: 1px solid #119fff !important;
      outline-offset: var(--theme-input--focus--outline-offset);
    }
  }

  .select.readonly {
    box-shadow: none !important;
  }

  .hidden {
    display: none !important;
  }

  .trigger {
    display: flex;
    align-items: center;
    flex-grow: 1;
    height: 100%;
  }

  .input-container {
    display: flex;
    position: relative;
    align-items: flex-start;
    width: 100%;

    .chips {
      position: relative;
      display: flex;
      align-items: center;
      flex-wrap: wrap;
      height: 100%;
      max-height: 3.5rem;
      flex-grow: 1;
      overflow-y: auto;

      > ix-filter-chip {
        margin: 0.1rem;
      }
    }

    input {
      @include ellipsis;
      @include text-default;
      background: transparent;
      height: 1.75rem;
      width: 100%;

      &,
      &:hover,
      &:focus-visible {
        border: none;
        outline: none;
      }

      &::placeholder {
        color: var(--theme-input-select-icon--color);
      }

      &.hide-placeholder::placeholder {
        opacity: 0;
      }
    }
  }

  .chevron-down-container {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 2rem;
    min-width: 2rem;
    height: 2rem;

    @include hover {
      color: var(--theme-color-dynamic--hover);
    }

    @include active {
      color: var(--theme-color-dynamic--hover);
    }
  }

  .editable {
    .chevron-down-container {
      @include hover {
        color: var(--theme-color-std-text);
        background-color: var(
          --theme-btn-invisible-secondary--background--hover
        );
      }

      @include active {
        color: var(--theme-color-std-text);
        background-color: var(
          --theme-btn-invisible-secondary--background--active
        );
      }
    }
  }

  .add-item {
    display: flex;
    justify-content: flex-start;
    align-items: center;
    position: relative;
    width: 100%;
  }

  .select-list-header {
    @include ellipsis;
    display: flex;
    align-items: center;
    height: $default-control-height;
    color: var(--theme-select-list-item-hint--color);
    margin: 0 $small-space 0 $x-large-space;
  }
}
