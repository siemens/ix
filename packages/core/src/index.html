<!--
SPDX-FileCopyrightText: 2023 Siemens AG

SPDX-License-Identifier: MIT
-->

<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0" />
  <title>Stencil Component Starter</title>

  <script type="module" src="/build/siemens-ix.esm.js"></script>
  <script nomodule src="/build/siemens-ix.js"></script>

  <link rel="stylesheet" href="/build/siemens-ix.css" />
  <link rel="stylesheet" href="/build/ix-brand-theme/dist/ix-brand-theme/ix-brand-theme.css" />
  <link rel="stylesheet" href="/build/ix-icons/css/ix-icons.css" />
  <script type="module">
    import { defineCustomElements } from '/build/ix-brand-theme/loader/index.es2017.js';
    defineCustomElements();
  </script>
</head>

<body style="margin: 0px; height: 100vh">
  <my-component />
</body>

</html>