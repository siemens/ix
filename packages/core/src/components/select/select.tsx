/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  Listen,
  Prop,
  State,
  Watch,
} from '@stencil/core';
import { IxSelectItemLabelChangeEvent } from '../select-item/events';

@Component({
  tag: 'ix-select',
  styleUrl: 'select.scss',
  scoped: true,
})
export class Select {
  @Element() hostElement!: HTMLIxSelectElement;

  /**
   * Indices of selected items
   * This corresponds to the value property of ix-select-items and therefor not neccessarily the indices of the items in the list.
   */
  @Prop({ mutable: true }) selectedIndices: string | string[] = [];

  /**
   * Show clear button
   */
  @Prop() allowClear = false;

  /**
   * Selection mode
   */
  @Prop() mode: 'single' | 'multiple' = 'single';

  /**
   * Select is extendable
   */
  @Prop() editable = false;

  /**
   * If true the select will be in disabled state
   */
  @Prop() disabled = false;

  /**
   * If true the select will be in readonly mode
   */
  @Prop() readonly = false;

  /**
   * Input field placeholder
   */
  @Prop() i18nPlaceholder = 'Select an option';

  /**
   * Input field placeholder for editable select
   */
  @Prop() i18nPlaceholderEditable = 'Type of select option';

  /**
   * Select list header
   */
  @Prop() i18nSelectListHeader = 'Please select an option';

  /**
   * Hint inside of dropdown if no items where found with current filter text
   *
   * @since 1.5.0
   */
  @Prop() i18nNoMatches = 'No matches';

  /**
   * Hide list header
   *
   * @since 1.5.0
   @
   */
  @Prop() hideListHeader = false;

  /**
   * Item selection changed
   */
  @Event() itemSelectionChange: EventEmitter<string | string[]>;

  /**
   * Item added to selection
   */
  @Event() addItem: EventEmitter<string>;

  @State() dropdownShow = false;
  @State() value: string[];
  @State() dropdownWrapperRef!: HTMLElement;
  @State() dropdownAnchor!: HTMLElement;
  @State() isDropdownEmpty = false;
  @State() hasFocus = false;
  @State() navigationItem: HTMLIxSelectItemElement;
  @State() inputFilterText: string;
  @State() inputValue: string;

  private inputRef!: HTMLInputElement;
  private dropdownRef!: HTMLIxDropdownElement;
  private addItemRef!: HTMLDivElement;

  private labelMutationObserver: MutationObserver;

  get items() {
    return Array.from(this.hostElement.querySelectorAll('ix-select-item'));
  }

  get selectedItems() {
    return this.items.filter((item) => item.selected);
  }

  get addItemButton() {
    return this.hostElement.querySelector('.add-item');
  }

  get isSingleMode() {
    return this.mode === 'single';
  }

  get isMultipleMode() {
    return this.mode === 'multiple';
  }

  get isEveryDropdownItemHidden() {
    return this.items.every((item) => item.classList.contains('d-none'));
  }

  @Watch('selectedIndices')
  watchSelectedIndices(newId: string | string[]) {
    if (!newId) {
      this.selectValue([]);
      return;
    }

    if (Array.isArray(newId)) {
      this.selectValue([...newId]);
      return;
    }

    this.selectValue([newId]);
  }

  @Listen('itemClick')
  onItemClicked(event: CustomEvent<string>) {
    const newId = event.detail;
    this.emitItemClick(newId);
  }

  private emitItemClick(newId: string) {
    if (this.isMultipleMode && Array.isArray(this.selectedIndices)) {
      if (this.selectedIndices.includes(newId)) {
        this.selectedIndices = this.selectedIndices.filter((i) => i !== newId);
      } else {
        this.selectedIndices = [...this.selectedIndices, newId];
      }
    } else {
      this.selectedIndices = [newId];
    }

    this.selectValue(this.selectedIndices);
    this.itemSelectionChange.emit(this.selectedIndices);
  }

  private emitAddItem(value: string) {
    if (value.trim() === '') {
      return;
    }

    const newItem = document.createElement('ix-select-item');
    newItem.value = value;
    newItem.label = value;

    this.addItemRef.appendChild(newItem);

    this.clearInput();
    this.emitItemClick(value);
    this.addItem.emit(value);
  }

  private selectValue(ids: string[]) {
    this.items.forEach((item) => {
      item.selected = ids.some((i) => i === item.value);
    });

    this.value = this.selectedItems.map((item) => item.label);

    if (this.isSingleMode) {
      this.inputValue = this.value?.length ? this.value[0] : null;
      return;
    }

    this.inputValue = null;
  }

  componentWillLoad() {
    if (this.selectedIndices) {
      this.selectValue(
        Array.isArray(this.selectedIndices)
          ? this.selectedIndices
          : [this.selectedIndices]
      );
    }
  }

  @Listen('ix-select-item:labelChange')
  onLabelChange(event: IxSelectItemLabelChangeEvent) {
    event.preventDefault();
    event.stopImmediatePropagation();
    this.selectValue(
      Array.isArray(this.selectedIndices)
        ? this.selectedIndices
        : [this.selectedIndices]
    );
  }

  disconnectedCallback() {
    if (this.labelMutationObserver) {
      this.labelMutationObserver.disconnect();
    }
  }

  private itemExists(item: string) {
    return this.items.find((i) => i.label === item);
  }

  private dropdownVisibilityChanged(event: CustomEvent<boolean>) {
    this.dropdownShow = event.detail;
    this.hasFocus = event.detail;

    if (event.detail) {
      this.inputRef.focus();
      this.inputRef.select();

      this.removeHiddenFromItems();
      this.isDropdownEmpty = this.isEveryDropdownItemHidden;
    } else {
      this.navigationItem = undefined;
    }
  }

  @Listen('keydown', {
    target: 'window',
  })
  async onKeyDown(event: KeyboardEvent) {
    if (!this.dropdownShow) {
      return;
    }

    if (event.code === 'ArrowDown' || event.code === 'ArrowUp') {
      this.onArrowNavigation(event);
    }

    if (event.code === 'Enter' || event.code === 'NumpadEnter') {
      await this.onEnterNavigation();
    }

    if (event.code === 'Escape') {
      this.dropdownShow = false;
    }
  }

  private async onEnterNavigation() {
    if (this.isMultipleMode) {
      return;
    }

    if (this.editable && !this.itemExists(this.inputFilterText)) {
      this.emitAddItem(this.inputFilterText);
      this.navigationItem = this.items[this.items.length - 1];
    }

    this.navigationItem?.onItemClick();
    await this.dropdownRef?.updatePosition();

    if (this.isSingleMode && !this.editable) {
      this.dropdownShow = false;
    }
  }

  private onArrowNavigation(event: KeyboardEvent) {
    event.stopPropagation();
    event.preventDefault();

    const focusItem = this.items.find(
      (item) => document.activeElement === item.querySelector('button')
    );
    this.navigationItem = focusItem;

    const selectItems = this.items.filter(
      (i) => !i.classList.contains('d-none')
    );

    const index = selectItems.indexOf(this.navigationItem);

    if (event.code === 'ArrowDown' && index < selectItems.length - 1) {
      this.navigationItem = selectItems[index + 1];
    } else if (event.code === 'ArrowUp' && index > 0) {
      this.navigationItem = selectItems[index - 1];
    }

    this.setHoverEffectForNavigatedSelectItem();
  }

  private setHoverEffectForNavigatedSelectItem() {
    this.navigationItem?.querySelector('button').focus();
  }

  private filterItemsWithTypeahead() {
    this.inputFilterText = this.inputRef.value;

    if (this.inputFilterText) {
      this.items.forEach((item) => {
        item.classList.remove('d-none');
        if (
          !item.label.toLowerCase().includes(this.inputFilterText.toLowerCase())
        ) {
          item.classList.add('d-none');
        }
      });
    } else {
      this.removeHiddenFromItems();
    }

    this.isDropdownEmpty = this.isEveryDropdownItemHidden;
  }

  private removeHiddenFromItems() {
    this.items.forEach((item) => {
      item.classList.remove('d-none');
    });
  }

  private clearInput() {
    this.inputRef.value = '';
    this.inputFilterText = '';
  }

  private clear() {
    this.clearInput();
    this.value = [];
    this.selectedIndices = [];
    this.itemSelectionChange.emit(null);
    this.dropdownShow = false;
  }

  private onInputBlur(e) {
    if (this.editable) {
      return;
    }

    if (this.isSingleMode) {
      if (this.dropdownShow && this.isDropdownEmpty) {
        this.dropdownShow = false;
      }
    }

    if (!this.dropdownShow && this.mode !== 'multiple') {
      e.target['value'] = this.value;
    }
  }

  private placeholderValue() {
    if (this.editable) {
      return this.i18nPlaceholderEditable;
    }

    if (this.readonly) {
      return '';
    }

    return this.i18nPlaceholder;
  }

  render() {
    return (
      <Host>
        <div
          class={{
            'form-control': true,
            select: true,
            focus: this.hasFocus,
            editable: this.editable,
            disabled: this.disabled,
            readonly: this.readonly,
          }}
          ref={(ref) => {
            this.dropdownAnchor = ref;
            if (!this.editable) this.dropdownWrapperRef = ref;
          }}
        >
          <div class="input-container">
            <div class="chips">
              {this.isMultipleMode
                ? this.selectedItems?.map((item) => (
                    <ix-filter-chip
                      disabled={this.disabled || this.readonly}
                      key={item.value}
                      onCloseClick={(e) => {
                        e.preventDefault();
                        e.stopPropagation();
                        this.emitItemClick(item.value);
                      }}
                    >
                      {item.label}
                    </ix-filter-chip>
                  ))
                : ''}
              <div class="trigger">
                <input
                  data-testid="input"
                  disabled={this.disabled}
                  readOnly={this.readonly}
                  type="text"
                  class={{
                    'allow-clear': this.allowClear && !!this.value?.length,
                  }}
                  placeholder={this.placeholderValue()}
                  value={this.inputValue}
                  ref={(ref) => (this.inputRef = ref)}
                  onBlur={(e) => this.onInputBlur(e)}
                  onInput={() => this.filterItemsWithTypeahead()}
                />
                {this.allowClear &&
                (this.value?.length || this.inputFilterText) ? (
                  <ix-icon-button
                    class="clear"
                    icon="clear"
                    ghost
                    oval
                    size="16"
                    onClick={(e) => {
                      e.preventDefault();
                      e.stopPropagation();
                      this.clear();
                    }}
                  />
                ) : null}
                {this.disabled || this.readonly ? null : (
                  <div
                    class="chevron-down-container"
                    ref={(ref) => {
                      if (this.editable) this.dropdownWrapperRef = ref;
                    }}
                  >
                    <ix-icon class="chevron" name="chevron-down-small" />
                  </div>
                )}
              </div>
            </div>
          </div>
        </div>
        <ix-dropdown
          ref={(ref) => (this.dropdownRef = ref)}
          show={this.dropdownShow}
          closeBehavior={this.isMultipleMode ? 'outside' : 'both'}
          class={{
            'd-none': this.disabled || this.readonly,
          }}
          anchor={this.dropdownAnchor}
          trigger={this.dropdownWrapperRef}
          onShowChanged={(e) => this.dropdownVisibilityChanged(e)}
          placement="auto-start"
          overwriteDropdownStyle={async () => {
            return {
              minWidth: `${this.hostElement.clientWidth}px`,
            };
          }}
        >
          <div
            class={{
              'select-list-header': true,
              hidden: this.hideListHeader || this.isDropdownEmpty,
            }}
            title={this.i18nSelectListHeader}
          >
            {this.i18nSelectListHeader}
          </div>
          <slot></slot>
          <div ref={(ref) => (this.addItemRef = ref)} class="d-contents"></div>
          {this.itemExists(this.inputFilterText) ? (
            ''
          ) : (
            <ix-dropdown-item
              data-testid="add-item"
              icon="plus"
              class={{
                'add-item': true,
                'd-none': !(this.editable && this.inputFilterText),
              }}
              label={this.inputFilterText}
              onItemClick={(e) => {
                e.preventDefault();
                e.stopPropagation();
                this.emitAddItem(this.inputFilterText);
              }}
            ></ix-dropdown-item>
          )}
          {this.isDropdownEmpty && !this.editable ? (
            <div class="select-list-header">{this.i18nNoMatches}</div>
          ) : (
            ''
          )}
        </ix-dropdown>
      </Host>
    );
  }
}
