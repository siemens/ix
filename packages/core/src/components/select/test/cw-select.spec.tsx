/*
 * SPDX-FileCopyrightText: 2022 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { newSpecPage } from '@stencil/core/testing';
import { fireEvent, screen } from '@testing-library/dom';
import { SelectItem } from '../../select-item/select-item';
import { Select } from '../select';

//@ts-ignore
global.MutationObserver = class {
  constructor() {}
  disconnect() {}
  observe() {}
};

describe('ix-select', () => {
  it('show add item button in list', async () => {
    const page = await newSpecPage({
      components: [Select, SelectItem],
      html: `<ix-select>
        <ix-select-item data-testid="select-1" value="11" label="Item 1"></ix-select-item>
        <ix-select-item data-testid="select-1" value="22" label="Item 2"></ix-select-item>
      </ix-select>`,
    });

    await page.waitForChanges();

    const selectElement = page.doc.querySelector('ix-select');

    selectElement.selectedIndices = '22';
    await page.waitForChanges();

    expect(
      (screen.getByTestId('input') as HTMLInputElement).value
    ).toStrictEqual('Item 2');
  });

  it('show add item button in list', async () => {
    const page = await newSpecPage({
      components: [Select],
      html: '<ix-select></ix-select>',
    });

    (page.win as any).SVGElement = class {};

    const select = page.doc.querySelector('ix-select');
    select.editable = true;
    await page.waitForChanges();

    const input = screen.getByTestId('input');

    const dropdownItem = screen.getByTestId('add-item');
    expect(dropdownItem.classList.contains('d-none')).toBeTruthy();
    const dropdown = page.root.querySelector('ix-dropdown');

    fireEvent(dropdown, new CustomEvent('showChanged'));
    await page.waitForChanges();

    fireEvent.input(input, { target: { value: '23' } });
    await page.waitForChanges();

    expect(dropdownItem).not.toHaveClass('d-none');
  });

  it('show not show add item button in list if editing is false', async () => {
    const page = await newSpecPage({
      components: [Select],
      html: '<ix-select></ix-select>',
    });

    (page.win as any).SVGElement = class {};

    const select = page.doc.querySelector('ix-select');
    select.editable = false;
    await page.waitForChanges();

    const input = screen.getByTestId('input');

    const dropdownItem = screen.getByTestId('add-item');
    expect(dropdownItem).toHaveClass('d-none');

    const dropdown = page.root.querySelector('ix-dropdown');

    fireEvent(dropdown, new CustomEvent('showChanged'));
    await page.waitForChanges();

    fireEvent.input(input, { target: { value: '23' } });
    await page.waitForChanges();

    expect(dropdownItem).toHaveClass('d-none');
  });

  it('show items', async () => {
    const page = await newSpecPage({
      components: [Select, SelectItem],
      html: `
      <ix-select>
        <ix-select-item data-testid="select-1" value="1" label="ABC"></ix-select-item>
        <ix-select-item data-testid="select-2" value="2" label="ABC 2"></ix-select-item>
        <ix-select-item data-testid="select-3" value="3" label="XYZ"></ix-select-item>
      </ix-select>`,
    });

    const dropdown = page.root.querySelector('ix-dropdown');
    fireEvent(dropdown, new CustomEvent('showChanged'));

    const input = screen.getByTestId('input');
    fireEvent.input(input, { target: { value: 'ABC' } });

    await page.waitForChanges();

    expect(screen.getByTestId('select-1')).not.toHaveClass('d-none');
    expect(screen.getByTestId('select-2')).not.toHaveClass('d-none');
    expect(screen.getByTestId('select-3')).toHaveClass('d-none');
  });

  describe('single mode selection', () => {
    it('change label of selected item', async () => {
      let updateCallback = jest.fn();

      //@ts-ignore
      global.MutationObserver = class {
        constructor(callback) {
          updateCallback = callback;
        }
        disconnect() {}
        observe() {}
      };

      const page = await newSpecPage({
        components: [Select, SelectItem],
        html: `
        <ix-select data-testid="select">
          <ix-select-item data-testid="select-1" value="1" label="ABC"></ix-select-item>
          <ix-select-item data-testid="select-2" value="2" label="ABC 2"></ix-select-item>
          <ix-select-item data-testid="select-3" value="3" label="XYZ"></ix-select-item>
        </ix-select>`,
      });
      await page.waitForChanges();

      const selectElement = page.doc.querySelector('ix-select');
      selectElement.selectedIndices = ['2'];
      await page.waitForChanges();

      const inputElement = screen.getByTestId('input') as HTMLInputElement;
      expect(inputElement.value).toBe('ABC 2');

      const selectItemElement = screen.getByTestId(
        'select-2'
      ) as HTMLIxSelectItemElement;
      selectItemElement.label = 'CHANGED!';
      updateCallback();

      await page.waitForChanges();

      expect(inputElement.value).toBe('CHANGED!');
    });

    it('select item', async () => {
      const page = await newSpecPage({
        components: [Select, SelectItem],
        html: `
        <ix-select data-testid="select">
          <ix-select-item data-testid="select-1" value="1" label="ABC"></ix-select-item>
          <ix-select-item data-testid="select-2" value="2" label="ABC 2"></ix-select-item>
          <ix-select-item data-testid="select-3" value="3" label="XYZ"></ix-select-item>
        </ix-select>`,
      });

      const selectionChange = jest.fn();

      screen
        .getByTestId('select')
        .addEventListener('itemSelectionChange', selectionChange);

      fireEvent(
        screen.getByTestId('select-2').querySelector('ix-dropdown-item'),
        new CustomEvent('itemClick')
      );
      await page.waitForChanges();

      expect((screen.getByTestId('input') as HTMLInputElement).value).toBe(
        'ABC 2'
      );
      expect(selectionChange).toHaveBeenCalled();
    });

    it('select new item', async () => {
      const page = await newSpecPage({
        components: [Select, SelectItem],
        html: `
        <ix-select data-testid="select">
          <ix-select-item data-testid="select-1" value="1" label="ABC"></ix-select-item>
          <ix-select-item data-testid="select-2" value="2" label="ABC 2"></ix-select-item>
          <ix-select-item data-testid="select-3" value="3" label="XYZ"></ix-select-item>
        </ix-select>`,
      });

      const selectElement: HTMLIxSelectElement = screen.getByTestId('select');
      selectElement.editable = true;

      await page.waitForChanges();

      const inputElement: HTMLInputElement = screen.getByTestId('input');
      fireEvent.input(inputElement, { target: { value: 'NEWITEM' } });

      await page.waitForChanges();

      fireEvent(screen.getByTestId('add-item'), new CustomEvent('itemClick'));

      await page.waitForChanges();

      expect(
        (
          page.root.querySelector(
            "ix-select-item[value='NEWITEM']"
          ) as HTMLIxSelectItemElement
        ).label
      ).toBe('NEWITEM');
    });

    it('select default', async () => {
      await newSpecPage({
        components: [Select, SelectItem],
        html: `
        <ix-select data-testid="select" selected-indices="2">
          <ix-select-item data-testid="select-1" value="1" label="ABC"></ix-select-item>
          <ix-select-item data-testid="select-2" value="2" label="ABC 2"></ix-select-item>
          <ix-select-item data-testid="select-3" value="3" label="XYZ"></ix-select-item>
        </ix-select>`,
      });

      expect(
        (screen.getByTestId('input') as HTMLInputElement).value
      ).toStrictEqual('ABC 2');
    });
  });

  describe('multiple mode selection', () => {
    it('select items', async () => {
      const page = await newSpecPage({
        components: [Select, SelectItem],
        html: `
        <ix-select data-testid="select" mode="multiple">
          <ix-select-item data-testid="select-1" value="1" label="ABC"></ix-select-item>
          <ix-select-item data-testid="select-2" value="2" label="ABC 2"></ix-select-item>
          <ix-select-item data-testid="select-3" value="3" label="XYZ"></ix-select-item>
        </ix-select>`,
      });

      const selectionChange = jest.fn();

      screen
        .getByTestId('select')
        .addEventListener('itemSelectionChange', selectionChange);

      fireEvent(
        screen.getByTestId('select-1').querySelector('ix-dropdown-item'),
        new CustomEvent('itemClick')
      );
      fireEvent(
        screen.getByTestId('select-3').querySelector('ix-dropdown-item'),
        new CustomEvent('itemClick')
      );
      await page.waitForChanges();

      expect(selectionChange.mock.calls[0][0].detail).toStrictEqual(['1']);
      expect(selectionChange.mock.calls[1][0].detail).toStrictEqual(['1', '3']);

      const chips = page.root.querySelectorAll('ix-filter-chip');

      expect(chips[0].innerText).toBe('ABC');
      expect(chips[1].innerText).toBe('XYZ');
    });

    it('select items and new items', async () => {
      const page = await newSpecPage({
        components: [Select, SelectItem],
        html: `
        <ix-select data-testid="select" mode="multiple">
          <ix-select-item data-testid="select-1" value="1" label="ABC"></ix-select-item>
          <ix-select-item data-testid="select-2" value="2" label="ABC 2"></ix-select-item>
          <ix-select-item data-testid="select-3" value="3" label="XYZ"></ix-select-item>
        </ix-select>`,
      });

      const selectionChangeHandle = jest.fn();
      const itemAddHandle = jest.fn();

      screen
        .getByTestId('select')
        .addEventListener('itemSelectionChange', selectionChangeHandle);
      screen.getByTestId('select').addEventListener('addItem', itemAddHandle);

      fireEvent(
        screen.getByTestId('select-1').querySelector('ix-dropdown-item'),
        new CustomEvent('itemClick')
      );
      fireEvent(
        screen.getByTestId('select-3').querySelector('ix-dropdown-item'),
        new CustomEvent('itemClick')
      );
      await page.waitForChanges();

      const dropdown = page.root.querySelector('ix-dropdown');
      fireEvent(dropdown, new CustomEvent('showChanged'));
      await page.waitForChanges();

      const input = screen.getByTestId('input');
      fireEvent.input(input, { target: { value: 'NEW ITEM' } });
      await page.waitForChanges();

      fireEvent(screen.getByTestId('add-item'), new CustomEvent('itemClick'));
      await page.waitForChanges();

      expect(itemAddHandle.mock.calls[0][0].detail).toStrictEqual('NEW ITEM');
      expect(selectionChangeHandle.mock.calls[2][0].detail).toStrictEqual([
        '1',
        '3',
        'NEW ITEM',
      ]);
      expect(
        (
          page.root.querySelector(
            "ix-select-item[value='NEW ITEM']"
          ) as HTMLIxSelectItemElement
        ).label
      ).toBe('NEW ITEM');
    });

    it('select default items', async () => {
      const page = await newSpecPage({
        components: [Select, SelectItem],
        html: `
        <ix-select data-testid="select" mode="multiple">
          <ix-select-item data-testid="select-1" value="1" label="ABC"></ix-select-item>
          <ix-select-item data-testid="select-2" value="2" label="ABC 2"></ix-select-item>
          <ix-select-item data-testid="select-3" value="3" label="XYZ"></ix-select-item>
        </ix-select>`,
      });

      const selectElement: HTMLIxSelectElement = screen.getByTestId('select');
      selectElement.selectedIndices = ['1', '3'];

      await page.waitForChanges();

      const chips = page.root.querySelectorAll('ix-filter-chip');

      expect(chips[0].innerText).toBe('ABC');
      expect(chips[1].innerText).toBe('XYZ');
    });
  });
});
