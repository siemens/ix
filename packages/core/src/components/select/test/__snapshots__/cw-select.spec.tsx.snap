// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ix-select renders 1`] = `
<ix-select>
  <!---->
  <div class="form-control select">
    <div class="input-container">
      <div class="chips">
        <div class="trigger">
          <input data-testid="input" placeholder="Select an option" type="text">
          <div class="chevron-down-container">
            <ix-icon class="chevron" name="chevron-down-small"></ix-icon>
          </div>
        </div>
      </div>
    </div>
  </div>
  <ix-dropdown adjustdropdownwidthtoreferencewidth="" placement="bottom" positioningstrategy="fixed" style="width: 100%;">
    <div class="select-list-header" title="Please select an option">
      Please select an option
    </div>
    <div class="d-contents"></div>
    <ix-dropdown-item class="add-item d-none" data-testid="add-item" icon="plus"></ix-dropdown-item>
  </ix-dropdown>
</ix-select>
`;
